### PR TITLE
Fixed Tiberian Sun units walking over icy water

### PR DESCRIPTION
**Pull request by Mailaender**
First steps towards https://github.com/OpenRA/OpenRA/issues/8918. This sets the initial state correctly. Also adds a `/debugice` chat command.

image

### CODE REVIEW

File: OpenRA.Mods.TS/OpenRA.Mods.TS.csproj [hunk 1]
```diff
@@ -69,14 +69,16 @@
     <Compile Include="Activities\VoxelHarvesterDockSequence.cs" />
     <Compile Include="SpriteLoaders\TmpTSLoader.cs" />
     <Compile Include="Traits\Buildings\TiberianSunRefinery.cs" />
+    <Compile Include="Traits\Render\RenderIceState.cs" />
     <Compile Include="Traits\Render\WithDockingOverlay.cs" />
     <Compile Include="Traits\Render\WithPermanentInjury.cs" />
     <Compile Include="Traits\Render\WithVoxelWalkerBody.cs" />
     <Compile Include="Traits\Render\WithVoxelUnloadBody.cs" />
     <Compile Include="Traits\Render\WithVoxelWaterBody.cs" />
-    <Compile Include="UtilityCommands\LegacyTilesetImporter.cs" />
```
Comment (GraionDilach) [inline OpenRA.Mods.TS/OpenRA.Mods.TS.csproj]: Whoops?

Comment (Mailaender) [inline OpenRA.Mods.TS/OpenRA.Mods.TS.csproj]: No, just resorting https://github.com/OpenRA/OpenRA/pull/10760/files#diff-c9b9ea950ff2d3777d20d61f3d92fbf9R81.

Comment (GraionDilach) [inline OpenRA.Mods.TS/OpenRA.Mods.TS.csproj]: Oh, ignore me then.

Comment (Mailaender) [inline OpenRA.Mods.TS/OpenRA.Mods.TS.csproj]: Put it into a separate commit now to clarify.

**Comment by abcdefg30** [comment]
Look good to me. :+1: / :white_check_mark:

File: OpenRA.Mods.TS/Traits/Render/RenderIceState.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+using System.Drawing;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Commands;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.TS.Traits
+{
+	[Desc("Displays ice strength above the tile.")]
+	class RenderIceStateInfo : ITraitInfo
```
Comment (pchote) [inline OpenRA.Mods.TS/Traits/Render/RenderIceState.cs]: Both traits live on the world actor, so it would be better to `Requires<IceLayer>` here.

File: OpenRA.Mods.TS/Traits/World/IceLayer.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+		};
+
+		public object Create(ActorInitializer init) { return new IceLayer(init.Self, this); }
+	}
+
+	class IceLayer : IWorldLoaded
+	{
+		readonly IceLayerInfo info;
+
+		public CellLayer<int> Strength;
```
Comment (pchote) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: You can initialize this in the constructor and then make it readonly.

**Comment by Mailaender** [comment]
Updated.

File: OpenRA.Mods.TS/Traits/World/IceLayer.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+				var template = w.TileSet.Templates[tile.Type];
+				if (info.StrengthPerTile.ContainsKey(template.Id))
+				{
+					var strength = info.StrengthPerTile[template.Id];
+					Strength[cell] = strength;
+
+					if (strength < 4)
+						w.Map.CustomTerrain[cell] = w.TileSet.GetTerrainIndex(info.ImpassableTerrainType);
+
+					// Saves some resources by only updating domain indices for previous water tiles.
```
Comment (pchote) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: What happens for strength = 2 or 3?

Comment (Mailaender) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: The initial state for all those tiles is `Clear` so they stay pathable.

Comment (pchote) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: In the line just above you explicitly set them to be impassable.

Comment (Mailaender) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: I am making the assumption that only the 100% water state (ice strength = 1) is relevant to separate larger land masses here. This saves us from checking all the decorational ice tiles until https://github.com/OpenRA/OpenRA/issues/10730 is solved and performance is not an issue anymore.

Comment (pchote) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: States 2 and 3 (what about 0?) separate the landmasses too, so this choice appears to leave you with inconsistent domains.

Comment (Mailaender) [inline OpenRA.Mods.TS/Traits/World/IceLayer.cs]: `0` is the cell layer's default value (no ice possible) which will be already taken care of at the initial domain index pass. I will set state 2 and 3 to passable for now.

**Comment by pchote** [comment]
As discussed in IRC, I believe the real issue here is with the assignment of strengths to tiles.  This will need to be rethought a bit.

**Comment by pchote** [comment]
IRC link for context: http://logs.openra.net/?year=2016&month=02&day=20#21:12:44

**Comment by reaperrr** [comment]
There hasn't been any progress in over two months, so I'm closing it for now to clear the PR queue.

**Comment by Mailaender** [comment]
I have no idea what to change here. Could simplify the assignments of strengths to

```csharp
enum Strength
{
	IcyWater = 1,
	CrackedIce = 2,
	Ice = 3
}
```

but that would then lose the information about neighbors and edges which I used to avoid glitches.
